### PR TITLE
De-traitify text module

### PR DESCRIPTION
**Pull request by PoignardAzur**

### CODE REVIEW

File: masonry/src/text/selection.rs [hunk 1]
```diff
@@ -675,27 +616,43 @@ impl<Str: Deref<Target = str> + TextStorage> Selectable for Str {
     }
 }
 
-/// A cursor type that implements `EditableTextCursor` for string types
+/// A cursor type with helper methods for working with strings.
 #[derive(Debug)]
 pub struct StringCursor<'a> {
```
Comment (DJMcNab) [inline masonry/src/text/selection.rs:621]: Why is this still `pub`? It has no public API except for construction.

File: masonry/src/text/selection.rs [hunk 2]
```diff
@@ -675,27 +616,43 @@ impl<Str: Deref<Target = str> + TextStorage> Selectable for Str {
...
 
-    fn is_boundary(&self) -> bool {
+    /// Check if cursor position is at a codepoint boundary.
+    pub(crate) fn is_boundary(&self) -> bool {
         self.text.is_char_boundary(self.position)
     }
 
-    fn prev(&mut self) -> Option<usize> {
+    /// Move cursor to previous codepoint boundary, if it exists.
+    /// Returns previous codepoint as usize offset.
```
Comment (DJMcNab) [inline masonry/src/text/selection.rs:654]:
```suggestion
    /// Move cursor to previous codepoint boundary, if it exists.
    /// Returns previous codepoint as usize offset, or `None` if this cursor was already at the first boundary.
```

File: masonry/src/widget/prose.rs [hunk 3]
```diff
@@ -226,6 +226,7 @@ impl Widget for Prose {
             }
             LifeCycle::BuildFocusChain => {
                 // TODO: This is *definitely* empty
+                #[cfg(FALSE)]
                 if !self.text_layout.text().links().is_empty() {
                     tracing::warn!("Links present in text, but not yet integrated");
                 }
```
Comment (DJMcNab) [inline masonry/src/widget/prose.rs:232]:
```suggestion
                // When we add links to `Prose`, they will probably need to be handled here.
```

File: masonry/src/widget/label.rs [hunk 4]
```diff
@@ -203,7 +203,9 @@ impl Widget for Label {
                 // TODO: Parley seems to require a relayout when colours change
                 ctx.request_layout();
             }
-            LifeCycle::BuildFocusChain => {
+            LifeCycle::BuildFocusChain =>
+            {
+                #[cfg(FALSE)]
                 if !self.text_layout.text().links().is_empty() {
                     tracing::warn!("Links present in text, but not yet integrated");
                 }
```
Comment (DJMcNab) [inline masonry/src/widget/label.rs:211]:
```suggestion
```
`Label` will never support links

**Review by DJMcNab — COMMENTED** [review]
I agree with this change. Not approving as PR is draft

**Comment by PoignardAzur** [comment]
Since I keep delaying the text refactor, and this is mostly ready, I'm removing the draft status and the TMP commit. I'll do the text refactor in another branch.

File: masonry/src/text/edit.rs [hunk 5]
```diff
@@ -364,49 +334,39 @@ impl<T: EditableText> TextEditor<T> {
...
+        let contents = self.text().clone();
         ctx.submit_action(Action::TextChanged(contents));
         Handled::Yes
     }
 }
 
-impl<T: EditableText> Deref for TextEditor<T> {
-    type Target = TextWithSelection<T>;
+impl Deref for TextEditor {
+    type Target = TextWithSelection<String>;
```
Comment (DJMcNab) [inline masonry/src/text/edit.rs:350]: I suspect we might want to remove this Deref at some point - it's a bit messy. Not a job for this PR

File: masonry/src/text/layout.rs [hunk 6]
```diff
@@ -475,15 +457,15 @@ impl<T: TextStorage> TextLayout<T> {
...
             builder.push_default(&StyleProperty::Brush(self.brush.clone()));
             builder.push_default(&StyleProperty::FontSize(self.text_size));
             builder.push_default(&StyleProperty::FontStack(self.font));
             builder.push_default(&StyleProperty::FontWeight(self.weight));
             builder.push_default(&StyleProperty::FontStyle(self.style));
-            // For more advanced features (e.g. variable font axes), these can be set in add_attributes
 
-            let builder = self.text.add_attributes(builder);
+            // Currently, this is used for underlining IME suggestions
+            // applying a brush to selected text.
```
Comment (DJMcNab) [inline masonry/src/text/layout.rs:468]:
```suggestion
            // Currently, this is used for:
            // - underlining IME suggestions
            // - applying a brush to selected text.
```
or
```suggestion
            // Currently, this is used for underlining IME suggestions
            // and applying a brush to selected text.
```

File: masonry/src/text/layout.rs [hunk 7]
```diff
@@ -520,10 +502,10 @@ impl<T: TextStorage> TextLayout<T> {
     }
 }
 
-impl<T: TextStorage> std::fmt::Debug for TextLayout<T> {
+impl<T: AsRef<str> + Eq> std::fmt::Debug for TextLayout<T> {
     fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
         f.debug_struct("TextLayout")
-            .field("text", &self.text.as_str().len())
+            .field("text", &self.text.as_ref())
```
Comment (DJMcNab) [inline masonry/src/text/layout.rs:508]: Maybe we want this to be truncated?

File: masonry/src/text/reference/movement2.rs
Comment (DJMcNab) [inline masonry/src/text/reference/movement2.rs:1]: ?

File: masonry/src/text/selection.rs [hunk 8]
```diff
@@ -541,58 +533,7 @@ pub trait Selectable: Sized + TextStorage {
...
-        };
-
-        if new_cursor.is_boundary() {
-            Some(new_cursor)
-        } else {
-            None
-        }
-    }
-
+impl<Str: Deref<Target = str> + AsRef<str> + Eq> Selectable for Str {
```
Comment (DJMcNab) [inline masonry/src/text/selection.rs:536]: I'm a bit confused why this needs both Deref and AsRef

File: masonry/src/widget/textbox.rs [hunk 9]
```diff
@@ -259,6 +259,7 @@ impl Widget for Textbox {
             LifeCycle::BuildFocusChain => {
                 ctx.register_for_focus();
                 // TODO: This will always be empty
+                #[cfg(FALSE)]
                 if !self.editor.text().links().is_empty() {
                     tracing::warn!("Links present in text, but not yet integrated");
                 }
```
Comment (DJMcNab) [inline masonry/src/widget/textbox.rs:265]: Probably should just have the same comment as `Prose`?

File: masonry/src/text/selection.rs [hunk 10]
```diff
@@ -686,27 +627,43 @@ impl<Str: Deref<Target = str> + TextStorage> Selectable for Str {
     }
 }
 
-/// A cursor type that implements `EditableTextCursor` for string types
+/// A cursor type with helper methods for working with strings.
```
Comment (DJMcNab) [inline masonry/src/text/selection.rs:630]:
```suggestion
/// A cursor type with helper methods for moving through strings.
```
This helps make it clear to readers that the methods are mutating a cursor position

**Review by DJMcNab — APPROVED** [review]
This is a big improvement, and doesn't seem to regress behaviour.

**Review by jaredoconnell — APPROVED** [review]
Tested without issue. It looks like an improvement. Should be good once Daniel's comments are addressed.

File: masonry/src/widget/textbox.rs (hunk 9 above)
Comment (PoignardAzur) [reply masonry/src/widget/textbox.rs]: I'm removing references to links for now. They were copy-pasted from Label's implementation, but I don't see links in editable text being implemented any time soon.

**Pull request merged**